File: examples/02-reliability/go/heartbeat_long_activity.go
package reliability

import (
	"context"
	"fmt"

	"go.temporal.io/sdk/activity"
)

// Heartbeat on every page. The heartbeat both reports progress (page becomes
// the resume point on retry, readable via activity.GetInfo) and is how a cancel
// is delivered: when the Workflow cancels, ctx.Done() fires here.
func ExportLargeTable(ctx context.Context, tableName string) (string, error) {
	for page := 0; page < 1000; page++ {
		select {
		case <-ctx.Done():
			cleanupPartialExport(tableName, page)
			return "", ctx.Err()
		default:
		}
		if err := exportPage(ctx, tableName, page); err != nil {
			return "", err
		}
		activity.RecordHeartbeat(ctx, page)
	}
	return fmt.Sprintf("s3://exports/%s", tableName), nil
}

File: examples/02-reliability/python/heartbeat_long_activity.py
import asyncio

from temporalio import activity


# Heartbeat on every page. The heartbeat both reports progress (page becomes the
# resume point on retry) and is how the server delivers a cancel: when the
# Workflow cancels, the next await point raises CancelledError here.
@activity.defn
async def export_large_table(table_name: str) -> str:
    for page in range(1000):
        try:
            await export_page(table_name, page)
            activity.heartbeat(page)
        except asyncio.CancelledError:
            await cleanup_partial_export(table_name, page)
            raise
    return f"s3://exports/{table_name}"
